app/services/bulk_downloads_validation_service.rb

```diff
@@ -0,0 +1,85 @@
+class BulkDownloadsValidationService
+  SAMPLE_ACCESS_ERROR = "Error validating samples. Please contact us for help.".freeze
+  UNKNOWN_DOWNLOAD_TYPE = "Could not find download type for bulk download".freeze
+  UPLOADER_ONLY_DOWNLOAD_TYPE = "You must be the uploader of all selected samples to initiate this download type.".freeze
+  MAX_SAMPLES_EXCEEDED_ERROR_TEMPLATE = "No more than %s samples allowed in one download.".freeze
+
+  def initialize(query_ids, current_user, max_samples_allowed)
+    @query_ids = query_ids.map(&:to_i)
+    @user = current_user
+    @max_samples = max_samples_allowed
+  end
+
+  def validate_sample_ids
+    error = nil
+
+    begin
+      viewable_samples = validate_sample_access(@query_ids, @user)
+      finalized_info = get_finalized_sample_ids(viewable_samples)
+    rescue => e
+      Rails.logger.warn("Error validating samples for bulk download while checking against max samples: #{e}")
+      error = e
+    end
+
+    result = if error.nil?
+               {
+                 valid_sample_ids: finalized_info[:valid_sample_ids],
+                 invalid_sample_names: finalized_info[:invalid_sample_names],
+                 error: error,
+               }
+             else
+               { valid_sample_ids: [], invalid_sample_names: [], error: error }
+             end
+
+    return result
+  end
+
+  private ### *** PRIVATE METHODS *** ###
+
+  # Returns only the samples viewable by the current user.
+  # Raises an error if more than the max samples allowed are passed.
+  def validate_sample_access(sample_ids, user)
+    # Max samples check.
+    max_samples_allowed = @max_samples
+
+    # Max samples should be string containing an integer, but just in case.
+    if max_samples_allowed.nil?
```

Comment:
I don't think there's a front-end check - I never encountered one even when I selected over a thousand samples. The front-end wouldn't know what max samples is set to anyways since it's configured in the backend right? But yeah there's no reason to check this twice so I'll get rid of it.

Reply:
There is a front-end check but it's disabled for admins. You can look at `handleBulkDownloadModalOpen` in `SamplesView`. The MAX_SAMPLES value is passed to the front-end via the UserContext object. (search for `user_context`)